### ISSUE TITLE
delay deamonize() until all animation files are loaded

### ISSUE THREAD

**Issue by tho-** [CoreELEC/splash-image#1]
In `init`, quickly after `load_splash()`, `/flash` is moved to `/sysroot/`.

When files are in `/splash` this is not an issue.
But when animation files are stored in `/flash/progress/`, the daemonize() call is done before all files are read. This leaves only a small time window when only one or two animations files can be loaded before `prepare_sysroot()` moves the mount point. The animation is thus greatly incomplete in this scenario.

The attached patch moves the daemonize() after all files are loaded. This should be enough to fix this issue, at the cost of delaying a bit the startup, though.

[delay.txt](https://github.com/CoreELEC/splash-image/files/6528017/delay.txt)

**Comment by Portisch** [comment]
I know this issue but had no solution for it.
The problem is the amount of RAM is needed. When now the demonize does happen after the reading of the frames it get read again. Each single frame does require 8MB on 1080p: 1920*1080*4bit.
So it will not show anything anymore on devices with low RAM (1-2GB). This was the reason I stopped research for it.

**Comment by tho-** [comment]
On linux, fork(2) uses copy-on-write semantics, so there is no memory
penalty after the fork. The child and parent share the same memory
pages.  Only when one of the two processes modify a page, the page is
duplicated. But since in splash-image the parent calls exit(3)
immediately, the fork(2) has basically no cost regarding memory.

So moving daemonize() around should have no impact on memory at all.
The images will be read by the parent into RAM, the parent will fork
and exit, and the images will be in the address space of the child at
no additional cost.

In addition, you could use mmap(2) instead of fread(3) to map each
image's contents in the virtual address space of the process. This
will consume even less RAM than the current version. (this is just the
same as what is done for mapping the frame buffer memory). If you are
interested in this, I can provide a patch in a separate new issue.

**Comment by Portisch** [comment]
Sure, please!
Maybe you just create a pull request as I can merge it easily and we get no author violence.

Thank you!

**Comment by vpeter4** [comment]
If I remember correctly there was one issue with fork using double amount of memory. Maybe because of the reason mentioned.

Also I think mmap would not help much here. The problem is that all the fullscreen images are keep in memory uncompressed and not just small part used for animation (to my knowledge).

**Comment by tho-** [comment]
You're actually right: mmap() will not help since all images data will
be read eventually.

As far as I can see in the code, what is kept in RAM is:
* all compressed images (same as on disk data)
* 2 copies of the uncompressed full base image (~16MB in 1080p)
* only one buffer for the current uncompressed animation image
  (normally small)

(Unless I read it wrong), the uncompressed animation buffer is
freed/realloc in each call of show_next_image(), so only one is active
at a time.

So I still wonder if there will be any memory issue if the daemonize()
is moved after reading all compressed image data. With images shipped
in the distribution it's 2MB only.

**Comment by vpeter4** [comment]
I don't remember anymore the details but this is _top_ output when current version runs on 2GB device. There is no change without and with your patch. Don't know why it doesn't use much more memory.
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
15867 root      20   0   37932  19540   1148 S  20.6   1.0   0:08.01 splash-image
```

**Comment by tho-** [comment]
Thanks for the info.

This makes sense: 16MB (2 uncompressed full size images) + 2.1M (raw
disk images) is roughly the 19MB resident memory.

With the patch, reading images from /flash should work again, then.

**Comment by Portisch** [comment]
I worked in past already on such solution. I can confirm reading the animation from `/flash` is working with this patch.
The problem in past was that the boot time was increased as the frames get read before the busybox init script is able to continue. All this is history related as the original splash tool source is, let's say "horrible" coded and occupied more than 2G of RAM for a 40 frame animation. This is the reason this new tool was introduced. To finish it up we only animate small peaces of the screen, not full screen anymore what will reduce the need of RAM memory even more.

So I checked it again with and without the patch and the memory consumption is same in both variants.
I will implement it a bit more intelligent way to check if animation is read from `/flash` or not and based on this determine the perform the daemonize before or after reading the frames.

Thank you for your support!
